hs_core/task_utils.py

```diff
@@ -28,15 +35,170 @@ def _retrieve_task_id(res_id, job_dict):
     return None
 
 
-def get_resource_bag_task(res_id):
+def _retrieve_job_id(job_name, res_id):
+    """
+    Retrieve a job id corresponding to the job_name and match the res_id
+    :param job_name: job name
+    :param res_id: resource id
+    :return: job id
+    """
     i = inspect()
     active_jobs = i.active()
-    job_id = _retrieve_task_id(res_id, active_jobs)
+    job_id = _retrieve_task_id(job_name, res_id, active_jobs)
     if not job_id:
         reserved_jobs = i.reserved()
-        job_id = _retrieve_task_id(res_id, reserved_jobs)
+        job_id = _retrieve_task_id(job_name, res_id, reserved_jobs)
         if not job_id:
             scheduled_jobs = i.scheduled()
-            job_id = _retrieve_task_id(res_id, scheduled_jobs)
-
+            job_id = _retrieve_task_id(job_name, res_id, scheduled_jobs)
     return job_id
+
+
+def get_or_create_task_notification(task_id, status='progress', name='', payload='', username='',
+                                    verify_task_id=False):
+    if verify_task_id:
+        filter_task = TaskNotification.objects.filter(task_id=task_id).first()
+        if filter_task:
+            task_dict = {
+                'id': task_id,
+                'name': filter_task.name,
+                'status': filter_task.status,
+                'payload': filter_task.payload
+            }
+            return task_dict
+        elif not task_exists(task_id):
```

Comment:
@hyi - I'm not following why we're not creating a task notification for anonymous users.  I don't see why we couldn't just follow the same workflow for anonymous users and just clear anonymous user TaskNotification entries after a day or two.  To take that further, I don't see why we couldn't clear TaskNotifications for all users after a day or two.  

Treating anonymous users differently as authenticated users within the workflows complicates the logic quite a bit and if we can treat all users as the same, we should.

Reply:
@sblack-usu I did it this way after careful thought. To support anonymous users, the endpoint such as get_task() cannot be protected with login required meaning bots/hackers could hit that endpoint repeatedly creating a large number of junk entries in our task notification model if we just use get_or_create to create task entries without differentiating between anonymous users and authenticated users. The bots/hackers could also "legitimately" download public resources repeatedly which inflates a large number of junk entries into our model as well. In addition, creating a clean up celery task to clean up entries would require timestamp comparison (hence requiring adding a new timestamp field) and set up an threshold to determine when a task entry should be cleared up, which is also more code and logic. Also I think the logic I added for handling this is mostly encapsulated in the get_or_create_task_notification function, so I think it should be OK.

Reply:
I would think extra TaskNotification entries are the least of our problems if bots/spam are repeatedly hitting the download endpoints.  Serving the downloads are much more likely to hurt performance.  A simple TaskNotification cleanup cron job probably should be implemented for authenticated users as well.

Reply:
Agreed serving the downloads are much more likely to hurt performance than the extra entries in the model, but it is still a problem which I think is better to avoid. I initially thought a cleanup cron job probably should be implemented for authenticated users as well, but I changed my mind after some more thought. Each authenticated user has a clean all icon to click to clear all notification entries associated with the user, so proactively clearing up notification entries on behalf of users does not seem like a good idea to me.